pkg/node/manager/manager.go

```diff
@@ -606,6 +619,16 @@ func (m *Manager) NodeDeleted(n nodeTypes.Node) {
 		return
 	}
 
+	remoteHostIdentity := identity.ReservedIdentityHost
```

Comment:
This block strikes me as too complex. It's at least difficult to read.. The name used is `remoteHostIdentity`, but it might end up as a plain numeric identity, or `host`. 

Can we remove this block and do this instead?

Pass the numeric identity to the call below
```
m.removeFromIDMD(ip, identity.NumericIdentity(n.NodeIdentity), n.Source)
```

Embed this logic of this block in a switch block within that function.

Reply:
Yeah it's complex, but it also matches the `NodeUpdated()` code to appropriately figure out the right host identity depending on how the agent flags for `--enable-remote-node-identity` are configured. We could refactor that logic to make a helper function to hide the complexity if that helps, but my concern with changing this logic is that we could inadvertently introduce another different bug by causing the upsert and delete functions to mismatch.

Reply:
I'll take a look at whether we could refactor / collapse these cases together to improve the readability and code re-use.

Reply:
How does this diff look? Functionally it should be the same but would make the various codepaths a little more consistent:

```
diff --git a/pkg/node/manager/manager.go b/pkg/node/manager/manager.go
index 60a51e464a16..1bf1aa8752b9 100644
--- a/pkg/node/manager/manager.go
+++ b/pkg/node/manager/manager.go
@@ -376,6 +376,19 @@ func (m *Manager) legacyNodeIpBehavior() bool {
        return true
 }
 
+func (m *Manager) calculateRemoteHostIdentity(nodeIdentity uint32) identity.NumericIdentity {
+       remoteHostIdentity := identity.ReservedIdentityHost
+       if m.conf.RemoteNodeIdentitiesEnabled() {
+               nid := identity.NumericIdentity(nodeIdentity)
+               if nid != identity.IdentityUnknown && nid != identity.ReservedIdentityHost {
+                       remoteHostIdentity = nid
+               } else if !nodeIdentity.IsLocal() {
+                       remoteHostIdentity = identity.ReservedIdentityRemoteNode
+               }
+       }
+       return remoteHostIdentity
+}
+
 // NodeUpdated is called after the information of a node has been updated. The
 // node in the manager is added or updated if the source is allowed to update
 // the node. If an update or addition has occurred, NodeUpdate() of the datapath
@@ -387,15 +400,7 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
        dpUpdate := true
        nodeIP := n.GetNodeIP(false)
 
-       remoteHostIdentity := identity.ReservedIdentityHost
-       if m.conf.RemoteNodeIdentitiesEnabled() {
-               nid := identity.NumericIdentity(n.NodeIdentity)
-               if nid != identity.IdentityUnknown && nid != identity.ReservedIdentityHost {
-                       remoteHostIdentity = nid
-               } else if !n.IsLocal() {
-                       remoteHostIdentity = identity.ReservedIdentityRemoteNode
-               }
-       }
+       remoteHostIdentity := m.calculateRemoteHostIdentity(n.NodeIdentity)
 
        var ipsAdded, healthIPsAdded, ingressIPsAdded []string
 
@@ -560,17 +565,25 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
 // (IDMD) map. The given node identity determines which labels are associated
 // with the CIDR.
 func (m *Manager) upsertIntoIDMD(prefix string, id identity.NumericIdentity) {
-       if id == identity.ReservedIdentityHost {
+       switch id {
+       case identity.IdentityUnknown:
+               // Explicit skip: Health and Ingress IPs aren't injected into
+               // m.ipcache.metadata currently.
+       case identity.ReservedIdentityHost:
                m.ipcache.UpsertMetadata(prefix, labels.LabelHost)
-       } else {
+       default:
                m.ipcache.UpsertMetadata(prefix, labels.LabelRemoteNode)
        }
 }
 
 func (m *Manager) removeFromIDMD(prefix string, id identity.NumericIdentity, src source.Source) {
-       if id == identity.ReservedIdentityHost {
+       switch id {
+       case identity.IdentityUnknown:
+               // Explicit skip: Health and Ingress IPs aren't injected into
+               // m.ipcache.metadata currently, so no need to remove them.
+       case identity.ReservedIdentityHost:
                m.ipcache.RemoveMetadata(prefix, labels.LabelHost, src)
-       } else {
+       default:
                m.ipcache.RemoveMetadata(prefix, labels.LabelRemoteNode, src)
        }
 }
@@ -619,16 +632,6 @@ func (m *Manager) NodeDeleted(n nodeTypes.Node) {
                return
        }
 
-       remoteHostIdentity := identity.ReservedIdentityHost
-       if m.conf.RemoteNodeIdentitiesEnabled() {
-               nid := identity.NumericIdentity(n.NodeIdentity)
-               if nid != identity.IdentityUnknown && nid != identity.ReservedIdentityHost {
-                       remoteHostIdentity = nid
-               } else if !n.IsLocal() {
-                       remoteHostIdentity = identity.ReservedIdentityRemoteNode
-               }
-       }
-
        // If the source is Kubernetes and the node is the node we are running on
        // Kubernetes is giving us a hint it is about to delete our node. Close down
        // the agent gracefully in this case.
@@ -653,6 +656,7 @@ func (m *Manager) NodeDeleted(n nodeTypes.Node) {
                        continue
                }
 
+               remoteHostIdentity := m.calculateRemoteHostIdentity(n.NodeIdentity)
                ip := address.IP.String()
                m.removeFromIDMD(ip, remoteHostIdentity, n.Source)
                m.ipcache.Delete(ip, n.Source)
```

Reply:
I looked at trying to shuffle this complexity into the removeIDMD() function but that results in making multiple unnecessary calls to the calculate function.

Ultimately though I think that the tidying up of this code would be better targeted at the main branch rather than in a bugfix. I don't think there is a functional difference here that we're exploring, and the code is pretty consistent with existing code (even if that code is a little ugly).

Reply:
Ah for that matter, the whole `upsertIntoIDMD` path is already cleaned up in v1.14/main, so there's no need for improving that. And the node identity labels function is already refactored.

Reply:
I definitely prefer this refactored version

Reply:
Saw your most recent two comments after reading and refreshing. I understand where you're coming from and I'll dismiss my review for this one.

Reply:
We may end up needing some additional refactoring like this sooner or later, depending on how critical the other bug mentioned [here](https://github.com/cilium/cilium/pull/27406#pullrequestreview-1572925607) ends up becoming. I'll store this up my sleeve..